components/script/dom/readablestream.rs

```diff
@@ -0,0 +1,517 @@
+    ) {
+        if available == 0 {
+            return;
+        }
+        unsafe {
+            let mut readable = false;
+            if !ReadableStreamIsReadable(*cx, stream, &mut readable) {
+                return;
+            }
+            if readable {
+                ReadableStreamUpdateDataAvailableFromSource(*cx, stream, available as u32);
+            }
+        }
+    }
+
+    /// Close a currently readable js stream.
+    #[allow(unsafe_code)]
+    fn maybe_close_js_stream(&self, cx: SafeJSContext, stream: HandleObject) {
+        unsafe {
+            let mut readable = false;
+            if !ReadableStreamIsReadable(*cx, stream, &mut readable) {
+                return;
+            }
+            if readable {
+                ReadableStreamClose(*cx, stream);
+            }
+        }
+    }
+
+    fn close(&self, cx: SafeJSContext, stream: HandleObject) {
+        *self.closed.borrow_mut() = true;
+        self.maybe_close_js_stream(cx, stream);
+    }
+
+    fn enqueue_chunk(&self, cx: SafeJSContext, stream: HandleObject, mut chunk: Vec<u8>) {
+        let available = {
+            let mut buffer = self.buffer.borrow_mut();
+            buffer.append(&mut chunk);
+            buffer.len()
+        };
+        self.maybe_signal_available_bytes(cx, stream, available);
+    }
+
+    #[allow(unsafe_code)]
+    fn pull(&self, cx: SafeJSContext, stream: HandleObject, _desired_size: usize) {
+        // Note: for pull sources,
+        // this would be the time to ask for a chunk.
+
+        if { *self.closed.borrow() } {
+            return self.maybe_close_js_stream(cx, stream);
+        }
+
+        let available = {
+            let buffer = self.buffer.borrow();
+            buffer.len()
+        };
+
+        self.maybe_signal_available_bytes(cx, stream, available);
+    }
+
+    fn get_chunk_with_length(&self, length: usize) -> Vec<u8> {
+        let mut buffer = self.buffer.borrow_mut();
+        let buffer_len = buffer.len();
+        assert!(buffer_len >= length as usize);
+        buffer.split_off(buffer_len - length)
+    }
+
+    #[allow(unsafe_code)]
+    fn write_into_buffer(
+        &self,
+        _cx: SafeJSContext,
+        _stream: HandleObject,
+        buffer: *mut c_void,
+        length: usize,
+        bytes_written: *mut usize,
+    ) {
+        let chunk = self.get_chunk_with_length(length);
+
+        unsafe {
+            *bytes_written = chunk.len();
+            ptr::copy_nonoverlapping(chunk.as_ptr(), buffer as *mut u8, chunk.len());
+        }
+    }
+}
+
+#[allow(unsafe_code)]
+/// Get the `done` property of an object that a read promise resolved to.
+pub fn get_read_promise_done(cx: SafeJSContext, v: &SafeHandleValue) -> Result<bool, Error> {
+    unsafe {
+        rooted!(in(*cx) let object = v.to_object());
+        rooted!(in(*cx) let mut done = UndefinedValue());
+        let has_done =
+            get_dictionary_property(*cx, object.handle(), "done", done.handle_mut()).is_ok();
```

Comment:
In the Err(()) case we need to deal with the pending JS exception, so we should return Error::JSFailed. Also `Ok(false)` means there was no `done` property present.